File: OsmAnd-java/src/main/java/net/osmand/router/RouteDataResources.java
package net.osmand.router;

import net.osmand.Location;
import net.osmand.binary.BinaryMapRouteReaderAdapter.RouteTypeRule;
import net.osmand.binary.RouteDataObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class RouteDataResources {

	private final Map<RouteTypeRule, Integer> rules = new LinkedHashMap<>();

	private final List<Location> locations;
	private int currentSegmentStartLocationIndex;

	private final List<Integer> routePointIndexes;

	private final Map<RouteDataObject, int[][]> pointNamesMap = new HashMap<>();

	public RouteDataResources() {
		this.locations = new ArrayList<>();
		routePointIndexes = new ArrayList<>();
	}

	public RouteDataResources(List<Location> locations, List<Integer> routePointIndexes) {
		this.locations = locations;
		this.routePointIndexes = routePointIndexes;
	}

	public Map<RouteTypeRule, Integer> getRules() {
		return rules;
	}

	public List<Location> getLocations() {
		return locations;
	}

	public List<Integer> getRoutePointIndexes() {
		return routePointIndexes;
	}

	public Location getCurrentSegmentLocation(int offset) {
		int locationIndex = currentSegmentStartLocationIndex + offset;
		if (locationIndex >= locations.size()) {
			throw new IllegalStateException("Locations index: " + locationIndex + " out of bounds");
		}
		return locations.get(locationIndex);
	}

	public int getCurrentSegmentStartLocationIndex() {
		return currentSegmentStartLocationIndex;
	}

	public void updateNextSegmentStartLocation(int currentSegmentLength) {
		int routePointIndex = routePointIndexes.indexOf(currentSegmentStartLocationIndex + currentSegmentLength);
		boolean overlappingNextRouteSegment =
				!(routePointIndex > 0 && routePointIndex < routePointIndexes.size() - 1);
		currentSegmentStartLocationIndex += overlappingNextRouteSegment
				? currentSegmentLength - 1
				: currentSegmentLength;
	}

	public Map<RouteDataObject, int[][]> getPointNamesMap() {
		return pointNamesMap;
	}
}

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/routing/RouteDataResources.kt
package net.osmand.shared.routing

import net.osmand.shared.data.KLocation
import kotlin.jvm.JvmOverloads

/**
 * The state shared by every segment of one route while it is written to or read from a gpx file.
 *
 * Segments are visited in order and each one consumes a run of [locations], so the reader has to
 * remember where the previous segment stopped. The encoding rules are collected here too, because
 * a gpx file carries one table of them for the whole route rather than one per segment.
 */
class RouteDataResources @JvmOverloads constructor(
	private val locations: MutableList<KLocation> = ArrayList(),
	private val routePointIndexes: MutableList<Int> = ArrayList()
) {

	private val rules: MutableMap<RouteTypeRule, Int> = LinkedHashMap()

	private val pointNamesMap: MutableMap<RouteDataObject, Array<IntArray?>> = HashMap()

	private var currentSegmentStartLocationIndex: Int = 0

	fun getRules(): MutableMap<RouteTypeRule, Int> = rules

	fun getLocations(): MutableList<KLocation> = locations

	fun getRoutePointIndexes(): MutableList<Int> = routePointIndexes

	fun getCurrentSegmentLocation(offset: Int): KLocation {
		val locationIndex = currentSegmentStartLocationIndex + offset
		if (locationIndex >= locations.size) {
			throw IllegalStateException("Locations index: $locationIndex out of bounds")
		}
		return locations[locationIndex]
	}

	fun getCurrentSegmentStartLocationIndex(): Int = currentSegmentStartLocationIndex

	/**
	 * Moves the cursor past the segment just handled. Consecutive segments share their meeting
	 * point, except where a route point falls on it, so the step is usually one short of the length.
	 */
	fun updateNextSegmentStartLocation(currentSegmentLength: Int) {
		val routePointIndex = routePointIndexes.indexOf(currentSegmentStartLocationIndex + currentSegmentLength)
		val overlappingNextRouteSegment = !(routePointIndex > 0 && routePointIndex < routePointIndexes.size - 1)
		currentSegmentStartLocationIndex += if (overlappingNextRouteSegment) {
			currentSegmentLength - 1
		} else {
			currentSegmentLength
		}
	}

	fun getPointNamesMap(): MutableMap<RouteDataObject, Array<IntArray?>> = pointNamesMap
}
